### PR TITLE
Implementation of Generic Datasource filter and Schema generation

### PR DESCRIPTION
**Pull request by IsaiasA1**
# Description

User can filter a datasource and generate a schema based on a model for the filter block.

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 80% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Tested locally via e2e

# Important!!!!

- Must update Functional Test for OS repository datasource since the input is now different
- Errors are no longer produced when an invalid or empty input is given, only returns an empty list

### CODE REVIEW

File: docs/data-sources/os_repository.md [hunk 1]
```diff
@@ -52,23 +52,26 @@ limitations under the License.
...
+#     id = ["1234","5678"]
+#   }
+# }
+
+# # Get OS Repository details by Name
+# data "powerflex_os_repository" "example3" {
+#   # this datasource supports filters like os repsoitory id, name, source path, etc.
+#   # Note: If both filters are used simultaneously, the results will include any records that match either of the filters.
+#   filter {
+#     in_use = false
```
Comment (doriac11) [inline docs/data-sources/os_repository.md:69]: Could you add an example for all fields that are possible to be filtered upon?

**Review by doriac11 — APPROVED** [review]

File: docs/data-sources/os_repository.md (hunk 1 above)
Comment (IsaiasA1) [reply docs/data-sources/os_repository.md:75]: done

**Comment by doriac11** [comment]
run e2e test

**Comment by baoy1** [comment]
Create job! Just curious user can write some HCL statement like below to get the "filtered" objects. What is the best benefit of the elaborate design?
```
locals {
  filtered_items = [
    for item in var.items : item 
    if item.some_property == some_value
  ]
}
```

**Comment by IsaiasA1** [comment]
> Create job! Just curious user can write some HCL statement like below to get the "filtered" objects. What is the best benefit of the elaborate design?
> 
> ```
> locals {
>   filtered_items = [
>     for item in var.items : item 
>     if item.some_property == some_value
>   ]
> }
> ```

This approach maybe a little more complicated depending on the number of filters being checked and requires extra understanding of HCL... although it's a little unpleasant looking the function call implemented outputs the same but without the drawbacks I mentioned. This was also addressed to Anupam before starting this story.

**Comment by IsaiasA1** [comment]
run e2e test

**Comment by doriac11** [comment]
run e2e test

**Review by shekhar-j — APPROVED** [review]
Nice work. Thanks!

**Comment by IsaiasA1** [comment]
run e2e test

File: powerflex/helper/helper.go [hunk 2]
```diff
@@ -295,3 +302,207 @@ func SetContains(set types.Set, value string) bool {
...
+
+		filteredArray, err = FilterByField(filteredArray, fieldValue, field)
+
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	allFilteredData := make([]interface{}, filteredArray.Len())
```
[the remark this answers is not in the document]
Comment (IsaiasA1) [reply powerflex/helper/helper.go:348]: you would have to convert it back regardless after wards from the filteredArray (reflect.value) into whatever interface{} is inputted. I tried just returning the reflect.value but it wouldn't recall the original value of interface{} input

**Comment by IsaiasA1** [comment]
run e2e test

**Pull request merged**